ArduSub/motors.cpp

```diff
@@ -130,11 +130,111 @@ void Sub::motors_output()
 {
     // check if we are performing the motor test
     if (ap.motor_test) {
-        return; // Placeholder
+        verify_motor_test();
+    } else {
+        motors.set_interlock(true);
+        motors.output();
     }
-    motors.set_interlock(true);
-    motors.output();
 }
+static uint32_t last_do_motor_test_fail_ms = 0;
+static uint32_t last_do_motor_test_ms = 0;
+// Initialize new style motor test
+// Perform checks to see if it is ok to begin the motor test
+// Returns true if motor test has begun
+bool Sub::init_motor_test()
+{
+    uint32_t tnow = AP_HAL::millis();
+
+    // Ten second cooldown period required with no do_set_motor requests required
+    // after failure.
+    if (tnow < last_do_motor_test_fail_ms + 10000 && last_do_motor_test_fail_ms > 0) {
+        gcs().send_text(MAV_SEVERITY_CRITICAL, "10 second cool down required");
+    }
+
+    // check if safety switch has been pushed
+    if (hal.util->safety_switch_state() == AP_HAL::Util::SAFETY_DISARMED) {
+        gcs().send_text(MAV_SEVERITY_CRITICAL,"Disarm hardware safety switch before testing motors.");
+        return false;
+    }
+
+    // Make sure we are on the ground
+    if (!motors.armed()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Arm motors before testing motors.");
+        return false;
+    }
+
+    enable_motor_output(); // set all motor outputs to zero
+    ap.motor_test = true;
+
+    return true;
+}
+
+// Verify new style motor test
+// The motor test will fail if the interval between received
+// MAV_CMD_DO_SET_MOTOR requests exceeds a timeout period
+// Returns true if it is ok to proceed with new style motor test
+bool Sub::verify_motor_test()
+{
+    bool pass = true;
+
+    // Require at least 2 Hz incoming do_set_motor requests
+    if (AP_HAL::millis() > last_do_motor_test_ms + 500) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Motor test timed out!");
+        pass = false;
+    }
+
+    if (!pass) {
+        ap.motor_test = false;
+        motors.armed(false); // disarm motors
+        last_do_motor_test_fail_ms = AP_HAL::millis();
+        return false;
+    }
+
+    return true;
+}
+
+bool Sub::handle_do_motor_test(mavlink_command_long_t command) {
+    last_do_motor_test_ms = AP_HAL::millis();
+
+    // If we are not already testing motors, initialize test
+    if(!ap.motor_test) {
+        if (!init_motor_test()) {
+            gcs().send_text(MAV_SEVERITY_WARNING, "motor test initialization failed!");
+            return false; // init fail
+        }
+    }
+
+    float motor_number = command.param1;
+    float throttle_type = command.param2;
+    float throttle = command.param3;
+    // float timeout_s = command.param4; // not used
+    float test_type = command.param5;
+
+    if (!is_equal(test_type, (float)MOTOR_TEST_ORDER_BOARD) &&
+        !is_equal(test_type, (float)MOTOR_TEST_ORDER_DEFAULT)) {
```

Comment:
Humm so default is BOARD? That's not what the other vehicles do. I think that having a common default to all ArduPilot vehicles would be best.

Reply:
This whole BOARD thing was introduced because the DEFAULT is useless for us. The other vehicles should default to board IMO to produce a consistent behavior.

https://github.com/ArduPilot/mavlink/pull/27

Reply:
I understand the DEFAULT is useless to you, but my point is precisely that we wouldn't need those new definitions if Sub was just going to have a different default order.

The other vehicles shouldn't change a behaviour that has been implemented for years. Sub is the new one, so following others is what makes sense - it's fine to not support the DEFAULT/SEQUENCE testing order, but I think it's wrong to change what the ArduPilot DEFAULT is for just one vehicle.

Reply:
So is the suggestion to reject both default/sequence in Sub, and accept only ORDER? It is still not clear to me what your problem is. I had it in mind as 'vehicle default' I think this is what was talked about but it was a year ago.

Reply:
Yes, my suggestion is to reject DEFAULT and SEQUENCE and only accept BOARD.

If it was for vehicle default, why did we need this enum? The default for Sub would be BOARD and for other vehicles it would be SEQUENCE, what was the issue?

My "problem" is trying to keep vehicles consistent, I see no reason for different vehicles to have different choices for what the default is.

Reply:
> If it was for vehicle default, why did we need this enum? The default for Sub would be BOARD and for other vehicles it would be SEQUENCE, what was the issue?

I don't remember, the entire process of trying to get this functionality into master was drawn out and confusing. I think people did not understand the idea of the UI we were trying to implement. I added it according to the review and recommendation of the other maintainers.

Reply:
> Yes, my suggestion is to reject DEFAULT and SEQUENCE and only accept BOARD.

Done.